### PR TITLE
provider-vsphere

### PR DESCRIPTION
**Pull request by MartinWeindel**
**What this PR does / why we need it**:
This PR contains the proof-of-concept infrastructure provider for VMware vSphere (with NSX-T).

- [x] Control plane controller
- [x] Control plane web hook
- [x]  Infrastructure controller
- [x]  Worker controller
- [x]  ~~etcd backup support~~ (Backup with MinIO will be handled in a separate PR)

Before we consider merge:
- [x] gardener/machine-controller-manager#384

### CODE REVIEW

**Comment by MartinWeindel** [comment]
Needs https://github.com/gardener/machine-controller-manager/pull/384

File: .gitignore [hunk 1]
```diff
@@ -3,6 +3,8 @@
 /dev
 /local
 **/dev
+/local
```
Comment (rfranzke) [inline .gitignore]: Already exists in line 4 :)

File: controllers/provider-vsphere/README.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+Please find more information regarding the extensibility concepts and a detailed proposal [here](https://github.com/gardener/gardener/blob/master/docs/proposals/01-extensibility.md).
+
+## Supported Kubernetes versions
+
+This extension controller supports the following Kubernetes versions:
+
+| Version         | Support     | Conformance test results |
+| --------------- | ----------- | ------------------------ |
+| Kubernetes 1.16 | 1.16.0+  ?  | -                        |
```
Comment (rfranzke) [inline controllers/provider-vsphere/README.md]: 1.17 is not supported yet?

Comment (MartinWeindel) [reply controllers/provider-vsphere/README.md]: 1.17 works too

File: controllers/provider-vsphere/charts/internal/shoot-system-components/charts/csi-vsphere/templates/csi-attacher-rbac.yaml [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+#   for non-namespaced entities like the ClusterRole
+# - decide whether the deployment replicates the external CSI
+#   attacher, in which case leadership election must be enabled;
+#   this influences the RBAC setup, see below
+
+# Attacher must be able to work with PVs, nodes and VolumeAttachments
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: garden.sapcloud.io:csi-attacher
```
Comment (rfranzke) [inline controllers/provider-vsphere/charts/internal/shoot-system-components/charts/csi-vsphere/templates/csi-attacher-rbac.yaml]: Can you please rename everything to `gardener.cloud`? We should not have `garden.sapcloud.io` anywhere anymore.

File: controllers/provider-vsphere/charts/internal/shoot-system-components/charts/vsphere-cloud-controller-manager/templates/rbac-cloud-controller-manager.yaml [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+      - secrets
+    verbs:
+      - get
+      - list
+      - watch
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: servicecatalog.k8s.io:apiserver-authentication-reader
```
Comment (rfranzke) [inline controllers/provider-vsphere/charts/internal/shoot-system-components/charts/vsphere-cloud-controller-manager/templates/rbac-cloud-controller-manager.yaml]: Service catalog?!

File: controllers/provider-vsphere/example/old/00-componentconfig.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
+---
```
Comment (rfranzke) [inline controllers/provider-vsphere/example/old/00-componentconfig.yaml]: Why are all these example files in an `old` subdirectory?

File: controllers/provider-vsphere/example/old/30-controlplane.yaml [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  apiToken: YXBpVG9rZW4K # apiToken
+  projectID: cHJvamVjdElkCg== # projectId
+---
+apiVersion: extensions.gardener.cloud/v1alpha1
+kind: Cluster
+metadata:
+  name: shoot--foobar--vsphere
+spec:
+  cloudProfile:
+    apiVersion: core.gardener.cloud/v1alpha1
```
Comment (rfranzke) [inline controllers/provider-vsphere/example/old/30-controlplane.yaml]: Can you change every occurrence of `core.gardener.cloud/v1alpha1` to `core.gardener.cloud/v1beta1`, please?

File: controllers/provider-vsphere/example/old/controller-registration.yaml [hunk 7]
```diff
@@ -0,0 +1,20 @@
+---
```
Comment (rfranzke) [inline controllers/provider-vsphere/example/old/controller-registration.yaml]: Can we delete this file? I don't think it's any helpful.

File: controllers/provider-vsphere/pkg/apis/vsphere/helper/helper_test.go [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+				Expect(err).To(HaveOccurred())
+			}
+		},
+
+		Entry("list is nil", nil, "ubuntu", "1", ""),
+		Entry("profile empty list", []api.MachineImages{}, "ubuntu", "1", ""),
+		Entry("profile entry not found (image does not exist)", makeProfileMachineImages("debian", "1"), "ubuntu", "1", ""),
+		Entry("profile entry not found (version does not exist)", makeProfileMachineImages("ubuntu", "2"), "ubuntu", "1", ""),
+		Entry("profile entry", makeProfileMachineImages("ubuntu", "1"), "ubuntu", "1", somePath),
+	)
```
Comment (rfranzke) [inline controllers/provider-vsphere/pkg/apis/vsphere/helper/helper_test.go]: Can we have a unit test for `CollectDatacenters` function?

File: controllers/provider-vsphere/pkg/apis/vsphere/validation/cloudprofile.go [hunk 9]
```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
```
Comment (rfranzke) [inline controllers/provider-vsphere/pkg/apis/vsphere/validation/cloudprofile.go]: Can we also have validation + tests for the `Infrastructure` and `ControlPlane` resources, please?

File: controllers/provider-vsphere/charts/internal/shoot-system-components/charts/csi-vsphere/templates/csi-attacher-rbac.yaml (hunk 3 above)
Comment (MartinWeindel) [reply controllers/provider-vsphere/charts/internal/shoot-system-components/charts/csi-vsphere/templates/csi-attacher-rbac.yaml]: done

File: controllers/provider-vsphere/example/old/00-componentconfig.yaml (hunk 5 above)
Comment (MartinWeindel) [reply controllers/provider-vsphere/example/old/00-componentconfig.yaml]: removed old directory

File: controllers/provider-vsphere/pkg/apis/vsphere/helper/helper_test.go (hunk 8 above)
Comment (MartinWeindel) [reply controllers/provider-vsphere/pkg/apis/vsphere/helper/helper_test.go]: done

File: controllers/provider-vsphere/pkg/apis/vsphere/validation/cloudprofile.go (hunk 9 above)
Comment (MartinWeindel) [reply controllers/provider-vsphere/pkg/apis/vsphere/validation/cloudprofile.go]: added test for ControlPlane. InfrastructureConfig is empty currently, i.e. there is nothing to test.

**Comment by rfranzke** [comment]
Can you rebase again to resolve the conflicts, please?

**Comment by rfranzke** [comment]
Once this PR is merged, can you work on another PR similar to https://github.com/gardener/gardener-extensions/pull/557 and https://github.com/gardener/gardener-extensions/pull/438?

File: controllers/provider-vsphere/charts/internal/machine-controller-manager/seed/templates/deployment.yaml [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        networking.gardener.cloud/to-seed-apiserver: allowed
+        networking.gardener.cloud/to-shoot-apiserver: allowed
+        networking.gardener.cloud/from-prometheus: allowed
+    spec:
+      serviceAccountName: machine-controller-manager
+      terminationGracePeriodSeconds: 5
+      containers:
+      - name: cmi-plugin
+        image: {{ index .Values.images "machine-controller-manager-provider-vsphere" }}
+        imagePullPolicy: Always
```
Comment (rfranzke) [inline controllers/provider-vsphere/charts/internal/machine-controller-manager/seed/templates/deployment.yaml]: This should be removed in favor of the line below.

File: controllers/provider-vsphere/charts/internal/shoot-system-components/charts/csi-vsphere/templates/vsphere-csi-syncer-rbac.yaml [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+    resources: ["persistentvolumes"]
+    verbs: ["get", "list", "watch"]
+---
+kind: ClusterRoleBinding
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: gardener.cloud:vsphere-csi-syncer
+subjects:
+  - kind: User
+    name: system:vsphere-csi-syncer
```
Comment (rfranzke) [inline controllers/provider-vsphere/charts/internal/shoot-system-components/charts/csi-vsphere/templates/vsphere-csi-syncer-rbac.yaml]:
```suggestion
    name: gardener.cloud:vsphere-csi-syncer
```

File: controllers/provider-vsphere/charts/provider-vsphere/values.yaml [hunk 12]
```diff
@@ -0,0 +1,61 @@
+image:
+  repository: eu.gcr.io/gardener-project/gardener/gardener-extension-hyper
+  tag: latest
+  pullPolicy: IfNotPresent
+
+replicaCount: 1
+resources: {}
```
Comment (rfranzke) [inline controllers/provider-vsphere/charts/provider-vsphere/values.yaml]: Can you define `vpa: true` to allow disabling the VPA in this chart?

File: controllers/provider-vsphere/pkg/apis/vsphere/types_cloudprofile.go [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+	TransportZone string
+	// LogicalTier0Router is the NSX-T logical tier 0 router
+	LogicalTier0Router string
+	// EdgeCluster is the NSX-T edge cluster
+	EdgeCluster string
+	// SNATIPPool is the NSX-T IP pool to allocate the SNAT ip address
+	SNATIPPool string
+
+	// Datacenter is the name of the vSphere data center (data center can either be defined at region or zone level)
+	// +optional
```
Comment (rfranzke) [inline controllers/provider-vsphere/pkg/apis/vsphere/types_cloudprofile.go]: Remove all `// +optional` tags from this file. Also, ensure that all optional fields are pointer types.

File: controllers/provider-vsphere/pkg/apis/vsphere/v1alpha1/types_cloudprofile.go [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+	// LogicalTier0Router is the NSX-T logical tier 0 router
+	LogicalTier0Router string `json:"logicalTier0Router"`
+	// EdgeCluster is the NSX-T edge cluster
+	EdgeCluster string `json:"edgeCluster"`
+	// SNATIPPool is the NSX-T IP pool to allocate the SNAT ip address
+	SNATIPPool string `json:"snatIPPool"`
+
+	// Datacenter is the name of the vSphere data center (data center can either be defined at region or zone level)
+	// +optional
+	Datacenter string `json:"datacenter,omitempty"`
```
Comment (rfranzke) [inline controllers/provider-vsphere/pkg/apis/vsphere/v1alpha1/types_cloudprofile.go]: See above: Ensure all optional fields have pointer types (please double-check all your resources, not only this file).

File: controllers/provider-vsphere/pkg/controller/controlplane/valuesprovider.go [hunk 15]
```diff
@@ -0,0 +1,567 @@
...
+				},
+				KubeConfigRequest: &secrets.KubeConfigRequest{
+					ClusterName:  clusterName,
+					APIServerURL: v1alpha1constants.DeploymentNameKubeAPIServer,
+				},
+			},
+			&secrets.ControlPlaneSecretConfig{
+				CertificateSecretConfig: &secrets.CertificateSecretConfig{
+					Name:         "vsphere-csi-syncer",
+					CommonName:   "system:vsphere-csi-syncer",
```
Comment (rfranzke) [inline controllers/provider-vsphere/pkg/controller/controlplane/valuesprovider.go]:
```suggestion
					CommonName:   "gardener.cloud:vsphere-csi-syncer",
```

File: controllers/provider-vsphere/pkg/controller/controlplane/valuesprovider.go [hunk 16]
```diff
@@ -0,0 +1,567 @@
...
+				},
+				KubeConfigRequest: &secrets.KubeConfigRequest{
+					ClusterName:  clusterName,
+					APIServerURL: v1alpha1constants.DeploymentNameKubeAPIServer,
+				},
+			},
+			&secrets.ControlPlaneSecretConfig{
+				CertificateSecretConfig: &secrets.CertificateSecretConfig{
+					Name:         "vsphere-csi-controller",
+					CommonName:   "system:vsphere-csi-controller",
```
Comment (rfranzke) [inline controllers/provider-vsphere/pkg/controller/controlplane/valuesprovider.go]:
```suggestion
					CommonName:   "gardener.cloud:vsphere-csi-controller",
```

File: controllers/provider-vsphere/charts/internal/shoot-system-components/charts/csi-vsphere/templates/vsphere-csi-controller-rbac.yaml [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+    resources: ["volumeattachments"]
+    verbs: ["get", "list", "watch", "update"]
+---
+kind: ClusterRoleBinding
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: gardener.cloud:vsphere-csi-controller
+subjects:
+  - kind: User
+    name: system:vsphere-csi-controller
```
Comment (rfranzke) [inline controllers/provider-vsphere/charts/internal/shoot-system-components/charts/csi-vsphere/templates/vsphere-csi-controller-rbac.yaml]:
```suggestion
    name: gardener.cloud:vsphere-csi-controller
```

File: controllers/provider-vsphere/pkg/controller/controlplane/valuesprovider.go [hunk 18]
```diff
@@ -0,0 +1,567 @@
...
+var controlPlaneShootChart = &chart.Chart{
+	Name: "shoot-system-components",
+	Path: filepath.Join(vsphere.InternalChartsPath, "shoot-system-components"),
+	SubCharts: []*chart.Chart{
+		{
+			Name: "vsphere-cloud-controller-manager",
+			Objects: []*chart.Object{
+				{Type: &corev1.ServiceAccount{}, Name: "cloud-controller-manager"},
+				{Type: &rbacv1.ClusterRole{}, Name: "system:cloud-controller-manager"},
+				{Type: &rbacv1.RoleBinding{}, Name: "servicecatalog.k8s.io:apiserver-authentication-reader"},
```
Comment (rfranzke) [inline controllers/provider-vsphere/pkg/controller/controlplane/valuesprovider.go]: Again, service catalog?

File: controllers/provider-vsphere/pkg/controller/controlplane/valuesprovider.go [hunk 19]
```diff
@@ -0,0 +1,567 @@
...
+	cluster *extensionscontroller.Cluster,
+) (map[string]interface{}, error) {
+
+	cloudProfileConfig, err := helper.GetCloudProfileConfig(&vp.ClientContext, cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	return map[string]interface{}{
+		"storagepolicyname": cloudProfileConfig.DefaultClassStoragePolicyName,
```
Comment (rfranzke) [inline controllers/provider-vsphere/pkg/controller/controlplane/valuesprovider.go]: CamelCase?

File: controllers/provider-vsphere/pkg/controller/worker/actuator.go [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+	)
+}
+
+type workerDelegate struct {
+	common.ClientContext
+
+	seedChartApplier gardener.ChartApplier
+	serverVersion    string
+
+	profileConfig *apisvsphere.CloudProfileConfig
```
Comment (rfranzke) [inline controllers/provider-vsphere/pkg/controller/worker/actuator.go]:
```suggestion
	cloudProfileConfig *apisvsphere.CloudProfileConfig
```

File: controllers/provider-vsphere/pkg/webhook/controlplane/ensurer.go [hunk 21]
```diff
@@ -0,0 +1,181 @@
...
+		ensureKubeControllerManagerCommandLineArgs(c)
+		ensureVolumeMounts(c)
+	}
+	ensureKubeControllerManagerAnnotations(template)
+	ensureVolumes(ps)
+	return e.ensureChecksumAnnotations(ctx, &dep.Spec.Template, dep.Namespace)
+}
+
+func ensureKubeAPIServerCommandLineArgs(c *corev1.Container) {
+	c.Command = extensionswebhook.EnsureStringWithPrefixContains(c.Command, "--enable-admission-plugins=",
```
Comment (rfranzke) [inline controllers/provider-vsphere/pkg/webhook/controlplane/ensurer.go]: Why do you enable this admission plugin? It doesn't make sense, you use CSI. Please remove it and rather ensure it's disabled.

File: controllers/provider-vsphere/pkg/webhook/controlplane/ensurer.go [hunk 22]
```diff
@@ -0,0 +1,181 @@
...
+	return e.ensureChecksumAnnotations(ctx, &dep.Spec.Template, dep.Namespace)
+}
+
+// EnsureKubeControllerManagerDeployment ensures that the kube-controller-manager deployment conforms to the provider requirements.
+func (e *ensurer) EnsureKubeControllerManagerDeployment(ctx context.Context, ectx genericmutator.EnsurerContext, dep *appsv1.Deployment) error {
+	template := &dep.Spec.Template
+	ps := &template.Spec
+	if c := extensionswebhook.ContainerWithName(ps.Containers, "kube-controller-manager"); c != nil {
+		ensureKubeControllerManagerCommandLineArgs(c)
+		ensureVolumeMounts(c)
```
Comment (rfranzke) [inline controllers/provider-vsphere/pkg/webhook/controlplane/ensurer.go]: Why do you mount the cloud provider information into the KCM? You use CCM and CSI, it doesn't make any sense. Or am I missing something?

File: controllers/provider-vsphere/pkg/webhook/controlplane/ensurer.go [hunk 23]
```diff
@@ -0,0 +1,181 @@
...
+}
+
+func ensureKubeControllerManagerAnnotations(t *corev1.PodTemplateSpec) {
+	t.Labels = extensionswebhook.EnsureAnnotationOrLabel(t.Labels, v1alpha1constants.LabelNetworkPolicyToPublicNetworks, v1alpha1constants.LabelNetworkPolicyAllowed)
+	t.Labels = extensionswebhook.EnsureAnnotationOrLabel(t.Labels, v1alpha1constants.LabelNetworkPolicyToPrivateNetworks, v1alpha1constants.LabelNetworkPolicyAllowed)
+	t.Labels = extensionswebhook.EnsureAnnotationOrLabel(t.Labels, v1alpha1constants.LabelNetworkPolicyToBlockedCIDRs, v1alpha1constants.LabelNetworkPolicyAllowed)
+}
+
+var (
+	cloudProviderConfigKubeControllerManagerVolumeMount = corev1.VolumeMount{
```
Comment (rfranzke) [inline controllers/provider-vsphere/pkg/webhook/controlplane/ensurer.go]: As far as I see all of this volume mount stuff can be removed as it's not needed for vSphere.

**Comment by vpnachev** [comment]
I hope our own development [guidelines](https://github.com/gardener/documentation/blob/master/CONTRIBUTING.md#pull-request-checklist) will be applied in the future.

**Pull request merged**